include/appimage/appimage_shared.h

```diff
@@ -12,8 +12,8 @@ extern "C" {
  */
 bool appimage_get_elf_section_offset_and_length(const char* fname, const char* section_name, unsigned long* offset, unsigned long* length);
 
-int appimage_print_hex(char* fname, unsigned long offset, unsigned long length);
-int appimage_print_binary(char* fname, unsigned long offset, unsigned long length);
+int appimage_print_hex(const char* fname, unsigned long offset, unsigned long length);
```

Comment:
A mutable pointer to a const char makes little sense in the context of C. If you are serious about constness, you should use `const char* const`.

Reply:
@TheAssassin sorry for late answer.

The data is immutable, that's the main point. So you can pass a string literal containing file name to that function.

Reply:
The data, as you pass it, is mutable, though. One cannot modify the pointer itself (read: location), but one can easily manipulate the memory at that pointer (read: contents). If you want to make sure a C string is not mutable, you need to pass `const` twice. See this example:

```
#include <string.h>
#include <stdio.h>

int main()
{
  const char* test = strdup("test");
  char* const test2 = strdup("test2");
  
  /*
    will break: 
     In function 'int main()':
    10:22: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
  */
  //strcpy(test, "abcd");
  
  // works
  test = test2;

  /*
    will not work:
     In function 'int main()':
    19:9: error: assignment of read-only variable 'test2'
    19:9: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
  */
  //test2 = test;
  
  // works
  strcpy(test2, "abcd");
  
  printf("test: %s\n", test);
  printf("test2: %s\n", test2);
  
  return 0;
}
```

In C, modifiers like `const` apply to the previous type specifier. For instance, in `char * const`, the `const` modifier applies to the pointer, i.e., it's a constant pointer (the value cannot be modified), but the memory at that location can still be mutated (e.g., you can call `strcpy` on it just fine). A `const char *` (or, better as in easier to understand, `char const *`) is a pointer to a constant `char`. You can modify the pointer's value (i.e., assign to the variable), but you cannot modify the contents (i.e., calling `strcpy` will yield a compiler error, as seen in the example above).

Therefore, if you really want to make this `const`, use `const` twice: `char const* const` (or `const char* const`).

The function does not manipulate the pointer in any way, and variables and parameters should be as `const` as possible anyway.

Reply:
@TheAssassin 

There is no need to be this rude. I know C. :)

Maybe code doesn't manipulate with address and indeed can be forced to be const, but this is useless information for callers. Nothing will change from it. In fact, `const` in C tells nothing to compiler and don't affect output assembly. It's only for humans to help them not make mistakes. I added this line only to fix a warning(which was forced to an error by compiler in my case) for existing code in tests: https://github.com/AppImage/libappimage/blob/master/tests/libappimage/legacy/test_shared.cpp#L70

Pushed an update with fixed `test_libappimage`, I didn't converted signed chars in the right way. By the way, it's not a good idea to use just chars to store non-characters. Its signness is implementation defined. Anyway, we're trying to teach each other how to write in C. :)